sdks/java/core/src/test/java/org/apache/beam/sdk/io/AvroIOTest.java

```diff
@@ -165,9 +165,105 @@ public GenericClass apply(GenericRecord input) {
     }
   }
 
+  private static final String SCHEMA_STRING =
+      "{\"namespace\": \"example.avro\",\n"
+          + " \"type\": \"record\",\n"
+          + " \"name\": \"AvroGeneratedUser\",\n"
+          + " \"fields\": [\n"
+          + "     {\"name\": \"name\", \"type\": \"string\"},\n"
+          + "     {\"name\": \"favorite_number\", \"type\": [\"int\", \"null\"]},\n"
+          + "     {\"name\": \"favorite_color\", \"type\": [\"string\", \"null\"]}\n"
+          + " ]\n"
+          + "}";
+
+  private static final Schema SCHEMA = new Schema.Parser().parse(SCHEMA_STRING);
+
   @Test
   @Category(NeedsRunner.class)
-  public void testAvroIOWriteAndReadAndParseASingleFile() throws Throwable {
+  public void testAvroIOWriteAndReadJavaClass() throws Throwable {
+    List<GenericClass> values =
+        ImmutableList.of(new GenericClass(3, "hi"), new GenericClass(5, "bar"));
+    File outputFile = tmpFolder.newFile("output.avro");
+
+    writePipeline
+        .apply(Create.of(values))
+        .apply(
+            AvroIO.write(GenericClass.class)
+                .to(writePipeline.newProvider(outputFile.getAbsolutePath()))
+                .withoutSharding());
+    writePipeline.run().waitUntilFinish();
+
+    PAssert.that(
+            readPipeline.apply(
+                "Read",
+                AvroIO.read(GenericClass.class)
+                    .from(readPipeline.newProvider(outputFile.getAbsolutePath()))))
+        .containsInAnyOrder(values);
+
+    readPipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testAvroIOWriteAndReadGeneratedClassWithSchema() throws Throwable {
+    File outputFile = tmpFolder.newFile("output.avro");
+
+    List<GenericRecord> values =
+        ImmutableList.<GenericRecord>of(
+            new AvroGeneratedUser("Bob", 256, null),
+            new AvroGeneratedUser("Alice", 128, null),
+            new AvroGeneratedUser("Ted", null, "white"));
+
+    writePipeline
+        .apply(Create.of(values))
+        .apply(
+            AvroIO.writeGenericRecords(SCHEMA)
+                .to(writePipeline.newProvider(outputFile.getAbsolutePath()))
+                .withoutSharding());
+    writePipeline.run().waitUntilFinish();
```

Comment:
Why is there sometimes a blocking call to `writePipeline.run().waitUntilFinish();` and sometimes a non blocking call to  `writePipeline.run()` for example l975 of `AvroIOTest`? In l975 for example, files written by the `writePipeline` are read just after the non blocking call to `writePipeline.run()`

Reply:
All waitUntilFinish() calls were unnecessary, I removed them.